# Movies!

**Because who doesn't love a good movie?**

We've all been there: it's another Friday night, and the COVID pandemic has limited our options for evening entertainment. Go to a party at a friend's house? Can't -- that'd be irresponsible and potentially dangerous. Check out a night club? Nope (let's be real, that was an awful option even before the pandemic hit).

And so, we're resorting once again to -- movie night! Make some stove top popcorn and plop down for 90 - 180 minutes of unbridled cinematic joy in the comfort of your own home (and pajamas).

But how can we possibly make the best choice possible when there are so many options? Nothing is worse than watching the ending credits flash on the screen and thinking (or exclaiming aloud, in extreme instances), "MAN, that sucked!" Here to help is a **movie recommender engine** using data obtained from <https://github.com/rashida048/Datasets/blob/master/movie_dataset.csv>

## Part I: Exploration

Let's read in the `dataset` and do some preliminary analysis.

In [1]:
# Make necessary imports
import pandas as pd
import numpy as np

In [2]:
# Read in the CSV file and check the head
movies = pd.read_csv('../datasets/movie_dataset.csv')
movies.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

In [3]:
# How many rows and columns does our dataset have?
movies.shape

(4803, 24)

In [4]:
# What are the names of the columns in our dataset?
movies.columns

Index(['index', 'budget', 'genres', 'homepage', 'id', 'keywords',
       'original_language', 'original_title', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director'],
      dtype='object')

In [5]:
# How many nulls are in our dataset?
movies.isnull().sum().sort_values(ascending = False)

homepage                3091
tagline                  844
keywords                 412
cast                      43
director                  30
genres                    28
overview                   3
runtime                    2
release_date               1
popularity                 0
budget                     0
id                         0
original_language          0
original_title             0
production_countries       0
production_companies       0
crew                       0
revenue                    0
spoken_languages           0
status                     0
title                      0
vote_average               0
vote_count                 0
index                      0
dtype: int64

In [6]:
# What datatype is each column?
movies.dtypes

index                     int64
budget                    int64
genres                   object
homepage                 object
id                        int64
keywords                 object
original_language        object
original_title           object
overview                 object
popularity              float64
production_companies     object
production_countries     object
release_date             object
revenue                   int64
runtime                 float64
spoken_languages         object
status                   object
tagline                  object
title                    object
vote_average            float64
vote_count                int64
cast                     object
crew                     object
director                 object
dtype: object

**Let's look at each column one by one to see what the data looks like**

**Index**

In [7]:
# Preview column index
movies[['index']]

index
0         0
1         1
2         2
3         3
4         4
...     ...
4798   4798
4799   4799
4800   4800
4801   4801
4802   4802

[4803 rows x 1 columns]

This data simply matches the already-existing index data of each row. This column can and should be `dropped`.

**Budget**

In [8]:
# Preview column budget
movies[['budget']]

budget
0     237000000
1     300000000
2     245000000
3     250000000
4     260000000
...         ...
4798     220000
4799       9000
4800          0
4801          0
4802          0

[4803 rows x 1 columns]

The data in this column is difficult to read due to the lack of `,` seperators. Let's assume the data is in American dollars as well.

**Genres**

In [9]:
# Preview column genres
movies[['genres']]

genres
0     Action Adventure Fantasy Science Fiction
1                     Adventure Fantasy Action
2                       Action Adventure Crime
3                  Action Crime Drama Thriller
4             Action Adventure Science Fiction
...                                        ...
4798                     Action Crime Thriller
4799                            Comedy Romance
4800             Comedy Drama Romance TV Movie
4801                                       NaN
4802                               Documentary

[4803 rows x 1 columns]

It looks like this column contains each genre that can be categorically applied to each movie with no sperator betweemn the genres (i.e., `,`). Let's look to see what the longest `genre` cell in this dataset look like.

In [10]:
# Sort movies by longest genre
movies.genres.str.len().sort_values(ascending = False)

1617    51.0
3208    50.0
305     50.0
1277    50.0
168     50.0
        ... 
4674     NaN
4681     NaN
4714     NaN
4716     NaN
4801     NaN
Name: genres, Length: 4803, dtype: float64

In [11]:
movies.iloc[1617]['genres']

'Action Adventure Animation Science Fiction Thriller'

This movie looks like the greatest one on the list! Who wouldn't want to watch an `Action/Adventure/Animation/Science Fiction/Thriller?`

When it comes time in this notebook for some `feature engineering`, we'll change this column into binarized columns per genre.

**Home page**

In [12]:
movies[['homepage']]

homepage
0                           http://www.avatarmovie.com/
1          http://disney.go.com/disneypictures/pirates/
2           http://www.sonypictures.com/movies/spectre/
3                    http://www.thedarkknightrises.com/
4                  http://movies.disney.com/john-carter
...                                                 ...
4798                                                NaN
4799                                                NaN
4800  http://www.hallmarkchannel.com/signedsealeddel...
4801                        http://shanghaicalling.com/
4802                                                NaN

[4803 rows x 1 columns]

This column isn't important for a recommender engine and can be dropped.

**ID**

In [13]:
movies[['id']]

id
0      19995
1        285
2     206647
3      49026
4      49529
...      ...
4798    9367
4799   72766
4800  231617
4801  126186
4802   25975

[4803 rows x 1 columns]

This column isn't important for a recommender engine and can be dropped.

**Keywords**

In [14]:
movies[['keywords']]

keywords
0     culture clash future space war space colony so...
1     ocean drug abuse exotic island east india trad...
2            spy based on novel secret agent sequel mi6
3     dc comics crime fighter terrorist secret ident...
4     based on novel mars medallion space travel pri...
...                                                 ...
4798  united states\u2013mexico barrier legs arms pa...
4799                                                NaN
4800  date love at first sight narration investigati...
4801                                                NaN
4802               obsession camcorder crush dream girl

[4803 rows x 1 columns]

This looks like a tricky column. Not only are there 412 `NaNs` in this column, but the keywords here look pretty subjective. Who decided on these keywords? The Academy at large? Due to its subjectivity and missing data, this column looks droppable from the final dataset.

**Original Language**

In [25]:
movies[['original_language']]

original_language
0                   en
1                   en
2                   en
3                   en
4                   en
...                ...
4798                es
4799                en
4800                en
4801                en
4802                en

[4803 rows x 1 columns]

In [24]:
# Check unique values of this column
movies.original_language.unique()

array(['en', 'ja', 'fr', 'zh', 'es', 'de', 'hi', 'ru', 'ko', 'te', 'cn',
       'it', 'nl', 'ta', 'sv', 'th', 'da', 'xx', 'hu', 'cs', 'pt', 'is',
       'tr', 'nb', 'af', 'pl', 'he', 'ar', 'vi', 'ky', 'id', 'ro', 'fa',
       'no', 'sl', 'ps', 'el'], dtype=object)

It's unclear what a majority of these abbreviations stand for. There's no key for what these might stand for on the website from which this dataset was pulled, which isn't helpful. Should  we drop this column from the dataset? Or attempt to find a key for this online somewhere? Probably not -- a different key might have been created by someone who would abbreviate languages differently, which could corrupt the accuracy of this dataset. 

In [33]:
movies.original_language.describe()

count     4803
unique      37
top         en
freq      4505
Name: original_language, dtype: object

It looks like there are 298 movies of the 4,803 in this dataset that aren't in English. In the interest of time, simplicity, and accuracy, it may be worth changing this project to a movie recommendation enginge in which all movies are in English.

In [35]:
# Look at the dataset for all non-English movies
movies[movies['original_language'] != 'en']

index    budget                                         genres  \
97       97  15000000  Action Adventure Drama Horror Science Fiction   
235     235  97250400                Fantasy Adventure Comedy Family   
317     317  94000000                              Drama History War   
474     474         0                           Mystery Drama Horror   
492     492   8000000                               Comedy Animation   
...     ...       ...                                            ...   
4739   4739         0                                 Comedy Romance   
4751   4751         0                                  Drama Romance   
4790   4790         0                                  Drama Foreign   
4792   4792     20000                  Crime Horror Mystery Thriller   
4798   4798    220000                          Action Crime Thriller   

                                               homepage      id  \
97                                                  NaN  315011   
235   http://www.asterixauxjeuxolympiques.com/index.php    2395   
317                http://www.theflowersofwarmovie.com/   76758   
474                                                 NaN  330770   
492                                                 NaN  293644   
...                                                 ...     ...   
4739                                                NaN   90414   
4751                                                NaN   42109   
4790                                                NaN   13898   
4792                                                NaN   36095   
4798                                                NaN    9367   

                                               keywords original_language  \
97     monster godzilla giant monster destruction kaiju                ja   
235   competition greece colosseum olympic games emp...                fr   
317                      forced prostitution child rape                zh   
474                        nurse sea beach boy pregnant                fr   
492                                                  3d                es   
...                                                 ...               ...   
4739                                     woman director                fr   
4751                                                NaN                pt   
4790                                                NaN                fa   
4792  japan prostitute hotel based on novel hallucin...                ja   
4798  united states\u2013mexico barrier legs arms pa...                es   

                          original_title  \
97                                シン・ゴジラ   
235          Astérix aux Jeux Olympiques   
317                                金陵十三釵   
474                            Évolution   
492   Don Gato: El inicio de la pandilla   
...                                  ...   
4739         I Love You, Don't Touch Me!   
4751            Gabriela, Cravo e Canela   
4790                               دایره   
4792                                 キュア   
4798                         El Mariachi   

                                               overview  popularity  ...  \
97    From the mind behind Evangelion comes a hit la...    9.476999  ...   
235   Astérix and Obélix have to win the Olympic Gam...   20.344364  ...   
317   A Westerner finds refuge with a group of women...   12.516546  ...   
474   11-year-old Nicolas lives with his mother in a...    3.300061  ...   
492   Top Cat has arrived to charm his way into your...    0.719996  ...   
...                                                 ...         ...  ...   
4739  The story of a 25 year old virgin girl, lookin...    0.020839  ...   
4751  In 1925, Gabriela becomes cook, mistress, and ...    0.557602  ...   
4790  Various women struggle to function in the oppr...    1.193779  ...   
4792  A wave of gruesome murders is sweeping Tokyo. ...    0.212443  ...   
4798  El Mariachi just wants to play his guita

After exploring this list on non-English movies further, my decision to drop these from the dataset is validated. Although the `overview` column for these movies are in English, what good is a movie recommendation for a film called "دایره" for someone who doesn't speak that native language? I searched for "دایره" on Google to see if I was able to find the movie described in this dataset, and I was unable to do so.   

**Original Title**

In [26]:
movies[['original_title']]

original_title
0                                       Avatar
1     Pirates of the Caribbean: At World's End
2                                      Spectre
3                        The Dark Knight Rises
4                                  John Carter
...                                        ...
4798                               El Mariachi
4799                                 Newlyweds
4800                 Signed, Sealed, Delivered
4801                          Shanghai Calling
4802                         My Date with Drew

[4803 rows x 1 columns]

In [30]:
movies[['overview']]

overview
0     In the 22nd century, a paraplegic Marine is di...
1     Captain Barbossa, long believed to be dead, ha...
2     A cryptic message from Bond’s past sends him o...
3     Following the death of District Attorney Harve...
4     John Carter is a war-weary, former military ca...
...                                                 ...
4798  El Mariachi just wants to play his guitar and ...
4799  A newlywed couple's honeymoon is upended by th...
4800  "Signed, Sealed, Delivered" introduces a dedic...
4801  When ambitious New York attorney Sam is sent t...
4802  Ever since the second grade when he first saw ...

[4803 rows x 1 columns]

This column looks more insightful than the `keywords` column. We can use `Natural Language Processing` when creating the model for this dataset.?

In [36]:
movies[['popularity']]

popularity
0     150.437577
1     139.082615
2     107.376788
3     112.312950
4      43.926995
...          ...
4798   14.269792
4799    0.642552
4800    1.444476
4801    0.857008
4802    1.929883

[4803 rows x 1 columns]

I don't like data columns like these. How is this measured? On what scale? And according to whom? This looks like something that can be dropped from our dataset.

In [38]:
movies[['production_companies']]

production_companies
0     [{"name": "Ingenious Film Partners", "id": 289...
1     [{"name": "Walt Disney Pictures", "id": 2}, {"...
2     [{"name": "Columbia Pictures", "id": 5}, {"nam...
3     [{"name": "Legendary Pictures", "id": 923}, {"...
4           [{"name": "Walt Disney Pictures", "id": 2}]
...                                                 ...
4798           [{"name": "Columbia Pictures", "id": 5}]
4799                                                 []
4800  [{"name": "Front Street Pictures", "id": 3958}...
4801                                                 []
4802  [{"name": "rusty bear entertainment", "id": 87...

[4803 rows x 1 columns]

In [51]:
(movies['production_companies'][0])

'[{"name": "Ingenious Film Partners", "id": 289}, {"name": "Twentieth Century Fox Film Corporation", "id": 306}, {"name": "Dune Entertainment", "id": 444}, {"name": "Lightstorm Entertainment", "id": 574}]'

In [55]:
(movies['production_companies'][0]).drop(0:2)

SyntaxError: invalid syntax (<ipython-input-55-04d70a5702fb>, line 1)